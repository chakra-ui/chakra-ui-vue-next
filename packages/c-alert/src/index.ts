import { defineComponent, computed, createVNode, mergeProps, isVNode, Fragment } from 'vue';
import { createStylesContext, useMultiStyleConfig, chakra } from '@chakra-ui/vue-system';
import { createContext, getValidChildren } from '@chakra-ui/vue-utils';
import { createIconComponent } from '@chakra-ui/c-icon';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };
  return _extends.apply(this, arguments);
}

var CInfoIcon = createIconComponent("info");
var CCheckIcon = createIconComponent("check-circle");
var CWarningIcon = createIconComponent("warning-alt");
var CErrorIcon = createIconComponent("warning");

function _isSlot(s) {
  return typeof s === 'function' || Object.prototype.toString.call(s) === '[object Object]' && !isVNode(s);
}

var STATUSES = {
  info: {
    colorScheme: "blue",
    icon: CInfoIcon
  },
  success: {
    colorScheme: "green",
    icon: CCheckIcon
  },
  warning: {
    colorScheme: "orange",
    icon: CWarningIcon
  },
  error: {
    colorScheme: "red",
    icon: CErrorIcon
  },
  loading: {
    icon: CInfoIcon,
    colorScheme: "blue"
  }
};

var _createStylesContext = createStylesContext("Alert"),
    StylesProvider = _createStylesContext[0],
    useStyles = _createStylesContext[1];

var _createContext = createContext({
  name: "AlertContext",
  errorMessage: "useAlertContext: `context` is undefined. Seems you forgot to wrap alert components in `<c-alert />`"
}),
    AlertProvider = _createContext[0],
    useAlertContext = _createContext[1];
/**
 * CAlert component
 *
 * This is the container component for all Alert components.
 * It also provides state and context to it's compound components
 */


var CAlert = defineComponent({
  name: "CAlert",
  props: {
    as: {
      type: [String, Object],
      "default": "div"
    },
    status: {
      type: [String],
      "default": "info"
    },
    colorScheme: {
      type: [String]
    },
    styleConfig: {
      type: [Object]
    },
    variant: {
      type: [String],
      "default": "solid"
    }
  },
  setup: function setup(props, _ref) {
    var slots = _ref.slots,
        attrs = _ref.attrs;
    var colorScheme = computed(function () {
      var _STATUSES$props$statu;

      return props.colorScheme || ((_STATUSES$props$statu = STATUSES[props == null ? void 0 : props.status]) == null ? void 0 : _STATUSES$props$statu.colorScheme);
    });
    var themingProps = computed(function () {
      return {
        colorScheme: colorScheme.value,
        variant: props.variant
      };
    });
    AlertProvider({
      status: computed(function () {
        return props.status;
      })
    });
    var styles = useMultiStyleConfig("Alert", themingProps);
    StylesProvider(styles);
    var alertStyles = computed(function () {
      return _extends({
        width: "100%",
        display: "flex",
        alignItems: "center",
        position: "relative",
        overflow: "hidden"
      }, styles.value.container);
    });
    return function () {
      return createVNode(chakra.div, mergeProps({
        "role": "alert",
        "__label": "alert",
        "__css": alertStyles.value
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertTitle component
 *
 * The title component for alerts
 */

var CAlertTitle = defineComponent({
  name: "CAlertTitle",
  setup: function setup(_, _ref2) {
    var attrs = _ref2.attrs,
        slots = _ref2.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__title",
        "__css": styles.value.title
      }, attrs), _isSlot(slots) ? slots : {
        "default": function _default() {
          return [slots];
        }
      });
    };
  }
});
/**Â«
 * CAlertDescription component
 *
 * The description component for alerts
 */

var CAlertDescription = defineComponent({
  name: "CAlertDescription",
  setup: function setup(_, _ref3) {
    var attrs = _ref3.attrs,
        slots = _ref3.slots;
    var styles = useStyles();
    return function () {
      return createVNode(chakra.div, mergeProps({
        "__label": "alert__description",
        "__css": styles.value.description
      }, attrs), {
        "default": function _default() {
          return getValidChildren(slots);
        }
      });
    };
  }
});
/**
 * CAlertIcon component
 *
 * The Icon component for alerts
 */

var CAlertIcon = defineComponent({
  name: "CAlertIcon",
  setup: function setup(_, _ref4) {
    var attrs = _ref4.attrs,
        slots = _ref4.slots;
    var styles = useStyles();

    var _useAlertContext = useAlertContext(),
        status = _useAlertContext.status;

    var BaseIcon = STATUSES[status.value].icon;
    var css = computed(function () {
      return status.value === "loading" ? styles.value.spinner : styles.value.icon;
    });
    return function () {
      var validChildren = getValidChildren(slots);
      return createVNode(chakra.span, mergeProps({
        "display": "inherit",
        "__label": "alert__icon"
      }, attrs, {
        "__css": css.value
      }), {
        "default": function _default() {
          return createVNode(Fragment, null, [validChildren.length ? slots : createVNode(BaseIcon, {
            "h": "100%",
            "w": "100%"
          }, null)]);
        }
      }); // return <icon {...styles.value.icon} {...attrs}></icon>
    };
  }
});

export { CAlert, CAlertDescription, CAlertIcon, CAlertTitle };
