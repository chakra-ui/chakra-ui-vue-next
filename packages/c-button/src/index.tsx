import { defineComponent } from "vue"

export { default as CButton } from "./button"
export { default as CButtonGroup } from "./button-group"
export { default as CIconButton } from "./icon-button"
