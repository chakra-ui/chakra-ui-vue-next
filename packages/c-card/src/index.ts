export { CCard, type CCardProps } from "./c-card"
export { CCardBody, type CCardBodyProps } from "./c-card-body"
export { CCardHeader, type CCardHeaderProps } from "./c-card-header"
