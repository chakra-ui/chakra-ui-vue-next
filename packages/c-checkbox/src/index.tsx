export { CCheckbox } from "./checkbox"
export {
  CCheckboxGroup,
  useCheckboxGroupContext,
  type CheckboxGroupContext,
} from "./checkbox-group"

export type { CCheckboxControlProps, CCheckboxProps } from "./checkbox"
export type { CCheckboxGroupProps } from "./checkbox-group"
