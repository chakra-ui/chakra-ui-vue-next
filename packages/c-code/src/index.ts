export { default as CCode } from './code'
