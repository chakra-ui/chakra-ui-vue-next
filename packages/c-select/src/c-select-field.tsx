import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"
import { PropType, SelectHTMLAttributes, defineComponent } from "vue"

type Omitted = "disabled" | "required" | "readOnly" | "size"

export interface SelectFieldProps
  extends Omit<HTMLChakraProps<SelectHTMLAttributes>, Omitted> {
  /**
   * @default false
   */
  disabled?: boolean
}

export const CSelectField = defineComponent({
  props: {
    placeholder: String as PropType<SelectFieldProps["placeholder"]>,
  },
  setup(props, { slots, attrs }) {
    return () => (
      <chakra.select __label="chakra-select" {...attrs}>
        {props.placeholder && <option value="">{props.placeholder}</option>}
        {slots.default?.()}
      </chakra.select>
    )
  },
})
