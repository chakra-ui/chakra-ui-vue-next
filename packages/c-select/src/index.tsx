export { CSelect, type CSelectProps } from "./c-select"
