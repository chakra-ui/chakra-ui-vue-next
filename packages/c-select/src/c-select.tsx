/**
 * Hey! Welcome to @chakra-ui/vue-next CSelect
 *
 * A component that allows users pick a value from predefined options
 *
 * @see Docs     https://next.vue.chakra-ui.com/c-select
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/main/packages/c-select/src/c-select/c-select.ts
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import {
  computed,
  defineComponent,
  h,
  mergeProps,
  PropType,
  reactive,
  SVGAttributes,
  ToRefs,
  toRefs,
  VNode,
} from "vue"
import { filterUndefined } from "@chakra-ui/utils"
import {
  chakra,
  HTMLChakraProps,
  omitThemingProps,
  SystemStyleObject,
  ThemingProps,
  useMultiStyleConfig,
} from "@chakra-ui/vue-system"
import {
  Assign,
  SAO,
  vueThemingProps,
  withSingleton,
} from "@chakra-ui/vue-utils"
import { FormControlOptions, useFormControl } from "@chakra-ui/c-form-control"

import { CSelectField, SelectFieldProps } from "./c-select-field"

interface RootProps extends Omit<HTMLChakraProps<"div">, "color"> {}

interface SelectOptions extends FormControlOptions {
  /**
   * The border color when the select is focused. Use color keys in `theme.colors`
   * @example
   * focusBorderColor = "blue.500"
   */
  focusBorderColor?: string
  /**
   * The border color when the select is invalid. Use color keys in `theme.colors`
   * @example
   * errorBorderColor = "red.500"
   */
  errorBorderColor?: string
  /**
   * The placeholder for the select. We render an `<option/>` element that has
   * empty value.
   *
   * ```jsx
   * <option value="">{placeholder}</option>
   * ```
   */
  placeholder?: string
  /**
   * The size (width and height) of the icon
   */
  iconSize?: string
  /**
   * The color of the icon
   */
  iconColor?: string
}

export interface CSelectProps
  extends SelectFieldProps,
    ThemingProps<"Select">,
    SelectOptions {
  /**
   * Props to forward to the root `div` element
   */
  rootProps?: RootProps
}

export const CSelect = defineComponent({
  name: "CSelect",
  props: {
    color: String,
    placeholder: String,
    iconColor: String as PropType<CSelectProps["iconColor"]>,
    isDisabled: Boolean as PropType<CSelectProps["isDisabled"]>,
    isInvalid: Boolean as PropType<CSelectProps["isInvalid"]>,
    focusBorderColor: SAO as PropType<CSelectProps["focusBorderColor"]>,
    errorBorderColor: SAO as PropType<CSelectProps["errorBorderColor"]>,
    ...vueThemingProps,
  },
  setup(props, { slots, attrs }) {
    const styles = useMultiStyleConfig("Select", props)

    const ownProps = computed(() => toRefs(reactive(omitThemingProps(props))))

    const fieldProps = useFormControl(ownProps.value)
    console.log(
      "ðŸš€ ~ file: c-select.tsx:104 ~ setup ~ fieldProps:",
      fieldProps.value
    )

    const rootStyles: SystemStyleObject = {
      width: "100%",
      height: "fit-content",
      position: "relative",
      color: props.color,
    }

    const fieldStyles: SystemStyleObject = {
      paddingEnd: "2rem",
      ...styles.value.field,
      _focus: {
        zIndex: "unset",
        ...(styles as any).value.field?.["_focus"],
      },
    }

    const iconColor = computed(() => props.iconColor ?? props.color)

    return () => (
      <chakra.div __css={rootStyles}>
        <CSelectField
          placeholder={props.placeholder}
          __css={fieldStyles}
          {...fieldProps.value}
          {...attrs}
        >
          {slots.default?.()}
        </CSelectField>
        <CSelectIcon
          data-disabled={fieldProps.value.disabled}
          {...fieldProps.value}
          color={iconColor.value}
          __css={styles.value.icon}
        />
      </chakra.div>
    )
  },
})

const CSelectIcon = defineComponent({
  name: "CSelectIcon",
  setup(_, { attrs }) {
    return () => (
      <chakra.div
        __label="select__icon-wrapper"
        position="absolute"
        display="inline-flex"
        alignItems="center"
        justifyContent="center"
        pointerEvents="none"
        top="50%"
        transform="translateY(-50%)"
        {...attrs}
      >
        <svg
          role="presentation"
          class="chakra-select__icon"
          viewBox="0 0 24 24"
          aria-hidden
          style={{
            width: "1em",
            height: "1em",
            color: "currentColor",
          }}
        >
          <path
            fill="currentColor"
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </chakra.div>
    )
  },
})
