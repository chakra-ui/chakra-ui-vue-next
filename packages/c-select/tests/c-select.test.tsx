import { Component } from "vue"
import { CFormControl } from "../../c-form-control/index"
import { render, testA11y } from "@chakra-ui/vue-test-utils"
import { CSelect } from "../src"

const renderComponent = (options?: Component) =>
  render({
    components: {
      CSelect,
    },
    template: `
      <CSelect aria-label="Select Food" placeholder="Select an option">
        <option value="a">A</option>
        <option value="b">B</option>
        <option value="c">C</option>
      </CSelect>`,
    ...options,
  })

describe("CSelect", () => {
  it("should render properly", () => {
    const { asFragment } = render(<CSelect />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("should pass a11y check", async () => {
    await testA11y(renderComponent())
  })

  it("renders a placeholder option", () => {
    const { container } = renderComponent()

    const option = container.querySelector("option[value='']") as HTMLElement

    expect(option).toBeInTheDocument()
    expect(option.textContent).toEqual("Select an option")
  })

  it("renders in disabled state if isDisabled is true", () => {
    const { container } = renderComponent({
      template: `<c-select isDisabled placeholder="Select an option" />`,
    })

    const select = container.querySelector("select") as HTMLElement
    const iconWrapper = container.querySelector(
      ".chakra-select__icon-wrapper"
    ) as HTMLElement

    expect(select).toBeDisabled()
    expect(iconWrapper).toHaveAttribute("data-disabled", "true")
  })

  it("doesn't render in disabled state if isDisabled is false", () => {
    const { container } = renderComponent({
      template: `<c-select :isDisabled="false" placeholder="Select an option" />`,
    })

    const select = container.querySelector("select") as HTMLElement
    const iconWrapper = container.querySelector(
      ".chakra-select__icon-wrapper"
    ) as HTMLElement

    expect(select).not.toBeDisabled()
    expect(iconWrapper).not.toHaveAttribute("data-disabled", "true")
  })

  it("renders in disabled state if wrapped by CFormControl with isDisabled true", () => {
    const { container, asFragment } = render({
      components: { CSelect, CFormControl },
      template: `
        <c-form-control is-disabled>
          <c-select placeholder="Select an option" />
        </c-form-control>
      `,
    })

    const select = container.querySelector("select") as HTMLElement
    const iconWrapper = container.querySelector(
      ".chakra-select__icon-wrapper"
    ) as HTMLElement

    expect(select).toBeDisabled()
    expect(iconWrapper).toHaveAttribute("data-disabled", "true")
    expect(asFragment()).toMatchSnapshot()
  })
})
