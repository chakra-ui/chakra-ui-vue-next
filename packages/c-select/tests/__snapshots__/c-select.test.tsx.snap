// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSelect renders in disabled state if wrapped by CFormControl with isDisabled true 1`] = `
<DocumentFragment>
  <div
    class="chakra-form chakra-1kxonj9"
    role="group"
  >
    <div
      class="chakra-42b2qy"
    >
      <select
        class="chakra-select chakra-1wd2nhn"
        disabled=""
        id="form"
        readonly="false"
      >
        <option
          value=""
        >
          Select an option
        </option>
        <!---->
      </select>
      <div
        class="chakra-select__icon-wrapper chakra-1hbwrp4"
        data-disabled="true"
        disabled="true"
        id="form"
        readonly="false"
        required="false"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CSelect should render properly 1`] = `
<DocumentFragment>
  <div
    class="chakra-42b2qy"
  >
    <select
      class="chakra-select chakra-1wd2nhn"
    >
      <!---->
      <!---->
    </select>
    <div
      class="chakra-select__icon-wrapper chakra-1hbwrp4"
    >
      <svg
        aria-hidden="true"
        class="chakra-select__icon"
        role="presentation"
        style="width: 1em; height: 1em; color: currentColor;"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
